File: twoSum/1679_MaxNumberOfKSumPairs.java
class Solution {
    public int maxOperations(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int count = 0;
        for (int i = 0; i < nums.length; i++) {
            int current = nums[i];
            int complement = k - current;
            if (map.getOrDefault(complement, 0) > 0) {
                // remove complement from the map
                map.put(complement, map.get(complement) - 1);
                count++;
            } else {
                 // add current element in the map
                map.put(current, map.getOrDefault(current, 0) + 1);
            }
        }
        return count;
    }
}

File: twoSum/1679_MaxNumberOfKSumPairs.py
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        num_ops = 0
        freq_count = defaultdict(int)
        for num in nums:
            if freq_count[k - num] > 0:
                freq_count[k - num] -= 1
                num_ops += 1
            else:
                freq_count[num] += 1
        return num_ops